File: JavaScript/0151. Reverse Words in a String.js
// Given an input string, reverse the string word by word.
//
// Example 1:
//
// Input: "the sky is blue"
// Output: "blue is sky the"
// Example 2:
//
// Input: "  hello world!  "
// Output: "world! hello"
// Explanation: Your reversed string should not contain leading or trailing spaces.
// Example 3:
//
// Input: "a good   example"
// Output: "example good a"
// Explanation: You need to reduce multiple spaces between two words to a single space in the reversed string.
//
// Note:
//
// A word is defined as a sequence of non-space characters.
// Input string may contain leading or trailing spaces. However, your reversed string should not contain leading or trailing spaces.
// You need to reduce multiple spaces between two words to a single space in the reversed string.
//
// Follow up:
//
// For C programmers, try to solve it in-place in O(1) extra space.

/**
 * @param {string} s
 * @return {string}
 */
const reverseWords = (s) => {
  return s.split(' ').reverse().join(' ').replace(/\s+/g, ' ').trim();
};

File: Python/0151. Reverse Words in a String.py
# Given an input string s, reverse the order of the words.
# A word is defined as a sequence of non-space characters. The words in s will be separated by at least one space.
# Return a string of the words in reverse order concatenated by a single space.
# Note that s may contain leading or trailing spaces or multiple spaces between two words. The returned string should only have a single space separating the words. Do not include any extra spaces.
#
# Example 1:
#
# Input: s = "the sky is blue"
# Output: "blue is sky the"
#
# Example 2:
#
# Input: s = "  hello world  "
# Output: "world hello"
# Explanation: Your reversed string should not contain leading or trailing spaces.
#
# Example 3:
#
# Input: s = "a good   example"
# Output: "example good a"
# Explanation: You need to reduce multiple spaces between two words to a single space in the reversed string.
#
# Constraints:
#
# 1 <= s.length <= 10^4
# s contains English letters (upper-case and lower-case), digits, and spaces ' '.
# There is at least one word in s.


class Solution:
    def reverseWords(self, s: str) -> str:
        # "a b  ".split(" ")  # ['a', 'b', '']
        # "a b  ".split()  # ['a', 'b']
        return " ".join(s.split()[::-1])
